finality-aleph/src/sync/ticker.rs

```diff
@@ -0,0 +1,122 @@
+use tokio::time::{sleep, Duration, Instant};
+
+/// This struct is used for rate limiting as an on-demand ticker. It can be used for ticking
+/// at least once `max_timeout` but not more than once every `min_timeout`.
+/// Example usage would be to use `wait` method in main select loop and `try_tick` whenever
+/// you would like to tick sooner in another branch of select.
+pub struct Ticker {
+    last_tick: Instant,
+    current_timeout: Duration,
+    max_timeout: Duration,
+    min_timeout: Duration,
+}
+
+impl Ticker {
+    /// Retruns new Ticker struct. Behaves as if last tick happened during creation of TIcker.
+    /// Requires `max_timeout` >= `min_timeout`.
+    pub fn new(mut max_timeout: Duration, min_timeout: Duration) -> Self {
+        if max_timeout < min_timeout {
+            max_timeout = min_timeout;
+        };
+        Self {
+            last_tick: Instant::now(),
+            current_timeout: max_timeout,
+            max_timeout,
+            min_timeout,
+        }
+    }
+
+    /// Returns whether at least `min_timeout` time elapsed since last tick.
+    /// If `min_timeout` elapsed since last tick, returns true and records a tick.
+    /// If not, returns false and calls to `wait` will return when `min_timeout`
+    /// elapses until the next tick.
```

Comment:
annoying nit: 
```suggestion
    /// Returns whether at least `min_timeout` time elapsed since the last tick.
    /// If `min_timeout` elapsed since the last tick, returns true and records a tick.
    /// If not, returns false and calls to `wait` will return when `min_timeout`
    /// elapses until the next tick.
```